### ISSUE TITLE
Better way to handle staging runtime methods

### ISSUE THREAD

**Issue by tdimitrov** [paritytech/polkadot#6714]
# What's the problem with the current versioning?
At the moment Versi and Rococo share a single runtime implementation. This means that if a new (staging) runtime api version is introduced it automatically appears in both networks.

The workflow for Versi is to have a 'stable' version of each runtime api. Each version after the stable one is considered staging. We are supposed to be free to add/remove/modify methods from the staging version freely and promote to stable once we decide to.

But there is a problem. If a Rococo runtime is upgraded while there are staging versions on Versi, they automatically go to Rococo and we can no longer handle them carelessly. This totally breaks the idea of staging runtime.

# Goals of the new versioning
what goals do we have?

1. Don't be required to release staging stuff, just because of some other emergency fix to be released ... this is something the minor version thing would also not automatically address.
2. Test version upgrades on test network as closely as it would be on a production network.
3. Be able to collect a few unstable things into a single to be release version.

# Possible solutions worth investigating
1. Have "hidden" minor versions which are only used on test networks. Then we can bump them all day long and once ready we bump the major version.
2. Separate Versi and Rococo code bases.
3. Introduce a cousin of `ParachainHost` runtime api and keep the staging stuff there. Once they are not staging - move them in `ParachainHost`.

**Comment by sandreim** [comment]
I think the cleanest way is 2. Are there any concerns with having a separate Versi Runtime,? this would also help with ParaThreads I think, as we do not want to have them running on Rococo while we are testing on Versi.
1 and 3 addd extra complexity and seem a bit hacky.

**Comment by eskimor** [comment]
A user story we need to cover somehow:

1. Developer develops new feature, which requires another runtime call.
2. Tests a runtime upgrade on Versi, enabling that feature/runtime call.
3. Does not work - needs to try again: How? Do we need to reset the network? - If that is a quick thing to do these days, might be fine, but it should be possible to iterate quickly and not wait a day to have another go for trying that typo fix.

Alternatively we improve our CI testing, so the above can be easily tested automatically without Versi even being involved.

**Comment by tdimitrov** [comment]
> I think the cleanest way is 2. Are there any concerns with having a separate Versi Runtime,?

Just more boilerplate code when adding a new runtime method, but at least in my opinion it's not a big deal. At this point I also vote for 2 :)

> Does not work - needs to try again: How? 

If he needs to modify the runtime call, he/she just does it. It's a test network so it doesn't matter if the logic is different, the parameters are different and etc. Some old request might fail, but who cares.

But if he wants to do an end-to-end test - then Versi needs to be reset. With the first deployment the storage was already modified and we can't undo this easily. The good thing is that runtime upgrades can be very well tested with `try-runtime` so I don't think someone will want to run this on Versi more than once twice.

**Comment by sandreim** [comment]
> A user story we need to cover somehow:
> 
> 1. Developer develops new feature, which requires another runtime call.
> 2. Tests a runtime upgrade on Versi, enabling that feature/runtime call.
> 3. Does not work - needs to try again: How? Do we need to reset the network? - If that is a quick thing to do these days, might be fine, but it should be possible to iterate quickly and not wait a day to have another go for trying that typo fix.
> 
> Alternatively we improve our CI testing, so the above can be easily tested automatically without Versi even being involved.

1, 2 are covered by separate runtimes. For 3 currently reseting Versi takes 30-40 minutes.

**Comment by sandreim** [comment]
> > I think the cleanest way is 2. Are there any concerns with having a separate Versi Runtime,?
> 
> Just more boilerplate code when adding a new runtime method, but at least in my opinion it's not a big deal. At this point I also vote for 2 :)
> 
Yeah, that is a good point, there is cost associated with having another runtime. 

> 
> But if he wants to do an end-to-end test - then Versi needs to be reset. With the first deployment the storage was already modified and we can't undo this easily. The good thing is that runtime upgrades can be very well tested with `try-runtime` so I don't think someone will want to run this on Versi more than once twice.

It is totally fine to break Versi since we can reset it fast.

**Comment by bkchr** [comment]
> Also regarding the 'try again' case. I think some initial testing can be performed locally? Zombienet still doesn't support runtime upgrades as part of a test, but one can spin up a network and upgrade it manually via polkadot.js?

Zombienet supports generic js scripts where you have access to the polkadot-js api and you can do all of these things.

In general, I don't understand why you are testing runtime upgrades on Versi. This doesn't make any sense. Versi also doesn't have any kind of state data attached that is worthwhile to have while running migrations or whatever. People can just run small test networks locally and destroy and rebuild them as often as they need. Versi should be for load testing etc, all the stuff that you can not simulate on your local computer. You can also write some Zombienet script which may helps you to retry the whole procedure so that in the end you just need to run some script.

In general to the versioning question, we could introduce features for `versi`. This is disabled by default and when enabled changes certain things in the Rococo runtime to make it Versi.
